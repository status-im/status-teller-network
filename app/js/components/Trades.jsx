import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col, Card } from 'reactstrap';
import { Link } from "react-router-dom";
import { withNamespaces } from 'react-i18next';
import Blockies from 'react-blockies';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faCircle } from "@fortawesome/free-solid-svg-icons";

class Trades extends Component {
  renderTrades() {
    return (
      <Card body className="py-0">
        {this.props.trades.map((trade, index) => (
          <div key={index} className="d-flex my-1">
            <span className="flex-fill align-self-center">
              <Blockies seed={trade.address} className="align-middle rounded-circle"/>
              <span className="ml-2">{trade.name}</span>
            </span>
            <span className="flex-fill align-self-center">{trade.value} ETH</span>
            <span className="flex-fill align-self-center text-right text-success">
              <FontAwesomeIcon icon={faCircle} className="mr-2"/>
              {trade.status}
            </span>
          </div>
        ))}
      </Card>
    );
  }

  renderEmpty(t) {
    return (
      <Card body className="text-center">
        {t('trades.noOpen')}
      </Card>
    );
  }

  render() {
    const t = this.props.t;
    return (
      <Row className="mt-4">
        <Col xs="12">
          <span className="font-weight-bold h5">{t('trades.title')}</span>
          <Link to="/buy" className="float-right">{t('trades.find')}</Link>
        </Col>
        <Col xs="12">
          {this.props.trades.length === 0 ? this.renderEmpty(t) : this.renderTrades()}
        </Col>
      </Row>
    );
  }
}

Trades.propTypes = {
  t: PropTypes.func,
  trades: PropTypes.array
};

export default withNamespaces()(Trades);
