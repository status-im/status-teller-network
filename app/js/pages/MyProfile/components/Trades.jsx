import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Card} from 'reactstrap';
import {Link} from "react-router-dom";
import {withNamespaces} from 'react-i18next';
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faCircle, faArrowRight} from "@fortawesome/free-solid-svg-icons";
import Identicon from "../../../components/UserInformation/Identicon";

class Trades extends Component {
  renderTrades() {
    return (
      <Card body className="py-2 px-3 shadow-sm">
        {this.props.trades.map((trade, index) => (
          <div key={index} className="d-flex my-1">
            <span className="flex-fill align-self-center">
              <Identicon seed={trade.offer.owner} className="align-middle rounded-circle border"/>
              <span className="ml-2">{trade.offer.user.username}</span>
            </span>
            <span className="flex-fill align-self-center">{trade.amount} {trade.offer.token.symbol}</span>
            <span className="flex-fill align-self-center text-right text-success">
              <FontAwesomeIcon icon={faCircle} className="mr-2"/>
              {trade.status}
            </span>
          </div>
        ))}
      </Card>
    );
  }

  renderEmpty() {
    const {t} = this.props;
    return (
      <Card body className="text-center">
        {t('trades.noOpen')}
      </Card>
    );
  }

  render() {
    const {t, trades} = this.props;
    return (
      <div className="mt-3">
        <div>
          <h3 className="d-inline-block">{t('trades.title')}</h3>
          <span className="float-right">
            <Link to="/buy" className="float-right">{t('trades.find')} <FontAwesomeIcon icon={faArrowRight}/></Link>
          </span>
        </div>
        {trades.length === 0 ? this.renderEmpty(t) : this.renderTrades()}
      </div>
    );
  }
}

Trades.propTypes = {
  t: PropTypes.func,
  trades: PropTypes.array
};

export default withNamespaces()(Trades);
