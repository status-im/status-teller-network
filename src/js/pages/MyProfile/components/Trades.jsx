import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {Card, Row, Col} from 'reactstrap';
import {Link} from "react-router-dom";
import {withNamespaces} from 'react-i18next';
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faArrowRight} from "@fortawesome/free-solid-svg-icons";
import Identicon from "../../../components/UserInformation/Identicon";
import {formatBalance} from "../../../utils/numbers";
import {tradeStates} from "../../../features/escrow/helpers";

const getTradeStyle = (trade) => {
  if(trade.arbitration){
    if(trade.arbitration.open){
      trade.status = tradeStates.arbitration_open;
    } else {
      if(trade.arbitration.result !== '0'){
        trade.status = tradeStates.arbitration_closed;
      }
    }
  }

  switch(trade.status){
    case tradeStates.waiting:
    case tradeStates.funded:
    case tradeStates.paid:
      return {text: 'Open', className: 'bg-success'};
    case tradeStates.released:
      return {text: 'Closed', className: 'bg-primary'};
    case tradeStates.canceled:
      return {text: 'Canceled', className: 'bg-secondary'};
    case tradeStates.expired:
      return {text: 'Expired', className: 'bg-secondary'};
    case tradeStates.arbitration_open:
      return {text: 'Arbitration', className: 'bg-danger'};
    case tradeStates.arbitration_closed:
      return {text: 'Resolved', className: 'bg-primary'};
    default:
      return {text: trade.status, className: 'bg-secondary'};
  }
};

class Trades extends Component {
  renderTrades() {
    const address = this.props.address;
    return (
      <Card body className="py-2 px-3 shadow-sm">
        {this.props.trades.map((trade, index) => {
          const isBuyer = trade.buyer ===  address;
          const tradeStyle = getTradeStyle(trade);
          return <Link key={index} to={"/escrow/" + trade.escrowId}>
            <Row className="my-1 border-bottom">
              <Col className="align-self-center pr-0" xs="2">
                <Identicon seed={ isBuyer ? trade.offer.owner : trade.buyer} scale={5} className="align-middle rounded-circle topCircle border"/>
              </Col>
              <Col className="align-self-center" xs="3">
                <span>{isBuyer ? trade.seller.username : trade.buyerInfo.username}</span>
              </Col>
              <Col className="align-self-center" xs="3">
                {isBuyer ? 'Buy' : 'Sell' } {formatBalance(trade.tokenAmount)} {trade.token.symbol}
              </Col>
              <Col className="align-self-center text-center text-success" xs="4">
                <span className={"p-1 text-uppercase d-inline text-white rounded-sm text-small " + tradeStyle.className}>{tradeStyle.text}</span>
              </Col>
            </Row>
          </Link>;
        })}
      </Card>
    );
  }

  renderEmpty() {
    const {t} = this.props;
    return (
      <Card body className="text-center">
        {t('trades.noOpen')}
      </Card>
    );
  }

  render() {
    const {t, trades} = this.props;
    return (
      <div className="mt-3">
        <div>
          <h3 className="d-inline-block">{t('trades.title')}</h3>
          <span className="float-right">
            <Link to="/buy" className="float-right">{t('trades.find')} <FontAwesomeIcon icon={faArrowRight}/></Link>
          </span>
        </div>
        {trades.length === 0 ? this.renderEmpty(t) : this.renderTrades()}
      </div>
    );
  }
}

Trades.propTypes = {
  t: PropTypes.func,
  trades: PropTypes.array,
  address: PropTypes.string
};

export default withNamespaces()(Trades);
