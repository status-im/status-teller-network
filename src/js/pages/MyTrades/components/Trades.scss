.profile-trades-list {
  overflow-y: auto;
  max-height: 270px;
}

.state-filters {
  .form-control {
    height: 40px;
  }
}
