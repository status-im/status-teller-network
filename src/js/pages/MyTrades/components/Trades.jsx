import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {Card, Row, Col, Form, FormGroup, Label, Input, CardBody, CardFooter} from 'reactstrap';
import {withNamespaces} from 'react-i18next';
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faCaretDown} from "@fortawesome/free-solid-svg-icons";
import {tradeStates, tradeStatesFormatted, completedStates} from "../../../features/escrow/helpers";
import {addressCompare} from "../../../utils/address";
import {truncateTwo} from '../../../utils/numbers';
import {calculateEscrowPrice} from '../../../utils/transaction';
import {ARBITRATION_SOLVED_BUYER, ARBITRATION_SOLVED_SELLER} from "../../../features/arbitration/constants";
import RoundedIcon from "../../../ui/RoundedIcon";
import classnames from 'classnames';
import './Trades.scss';
import {getTokenImage} from "../../../utils/images";
import messageImage from '../../../../images/read-chat.svg';
import moneyImage from '../../../../images/money-hand.svg';

import {PAYMENT_METHODS} from "../../../features/metadata/constants";

import UserInfoRow from "../../../components/UserInfoRow";

const getTradeStyle = (trade, isBuyer) => {
  if (trade.mining) {
    return {text: 'Mining', className: 'bg-info'};
  }
  if (trade.arbitration) {
    if (trade.arbitration.open) {
      trade.status = tradeStates.arbitration_open;
    } else {
      if (trade.arbitration.result !== '0') {
        trade.status = tradeStates.arbitration_closed;
      }
    }
  }
  const tradeStyle = {text: tradeStatesFormatted[trade.status]};

  const actionNeeded = 'Action needed';
  const sellersTurn = "Seller's turn";
  const buyersTurn = "Buyer's turn";

  switch (trade.status) {
    case tradeStates.waiting:
      tradeStyle.className = isBuyer ? 'bg-dark' : 'bg-warning';
      tradeStyle.text = isBuyer ? sellersTurn : actionNeeded;
      break;
    case tradeStates.funded:
      tradeStyle.className = isBuyer ? 'bg-warning' : 'bg-dark';
      tradeStyle.text = isBuyer ? actionNeeded : buyersTurn;
      break;
    case tradeStates.paid:
      tradeStyle.className = isBuyer ? 'bg-dark' : 'bg-warning';
      tradeStyle.text = isBuyer ? sellersTurn : actionNeeded;
      break;
    case tradeStates.released:
      tradeStyle.className = 'bg-success';
      break;
    case tradeStates.canceled:
      tradeStyle.className = 'bg-danger';
      break;
    case tradeStates.expired:
      tradeStyle.className = 'bg-secondary text-black';
      break;
    case tradeStates.arbitration_open: tradeStyle.className = 'bg-warning'; break;
    case tradeStates.arbitration_closed: {
      let className;
      if (trade.arbitration.result.toString() === ARBITRATION_SOLVED_BUYER) {
        className = isBuyer ? 'bg-success' : 'bg-danger';
      } else if (trade.arbitration.result.toString() === ARBITRATION_SOLVED_SELLER) {
        className = !isBuyer ? 'bg-success' : 'bg-danger';
      } else {
        className = 'bg-primary';
      }
      tradeStyle.className = className; break;
    }
    default: tradeStyle.className = 'bg-secondary text-black'; break;
  }
  return tradeStyle;
};

class Trades extends Component {
  state = {
    filteredState: '',
    showFilters: false
  };

  filterState(stateName) {
    if (stateName === '') {
      return this.setState({filteredState: ''});
    }
    const stateIndex = Object.values(tradeStatesFormatted).findIndex(tradeState => tradeState === stateName);
    const filteredState = stateIndex === -1 ? '' : Object.keys(tradeStatesFormatted)[stateIndex];
    this.setState({filteredState});
  }

  renderTrades() {
    return (
      <Fragment>
        <div className="profile-trades-list border-0 pt-0">
          {(() => {
            const trades = this.props.trades.filter(x => { return this.props.active ? !completedStates.includes(x.status) : completedStates.includes(x.status); }).map((trade, index) => {
              if (this.state.filteredState && trade.status !== this.state.filteredState) {
                return null;
              }

              const isBuyer = addressCompare(trade.buyer, this.props.address);
              const tradeStyle = getTradeStyle(trade, isBuyer);

              return (<Card key={index} className={classnames("clickable mb-3 shadow border-0 offer-card", {"card-transparent": !this.props.active})}
                            onClick={() => this.props.tradeClick(trade.escrowId)}>
                <CardBody>
                  <UserInfoRow hideAddress user={isBuyer ? trade.seller : trade.buyerInfo}
                               address={isBuyer ? trade.buyer : trade.offer.owner}
                               lastColSize={4}
                               lastCol={<div className="text-right">
                                 <span
                                   className={"p-1 px-2 d-inline text-white rounded-sm text-small nowrap " + tradeStyle.className}>{tradeStyle.text}</span>
                               </div>}/>
                   <Row className="mt-2">
                     <Col xs={2}><RoundedIcon image={messageImage} bgColor="blue" size="sm"/></Col>
                     <Col xs={10} className="pl-0">Status {/*TODO Put the real comm method once contract is implemented*/}</Col>
                   </Row>
                   <Row className="mt-2">
                     <Col xs={2}><RoundedIcon image={moneyImage} bgColor="blue" size="sm"/></Col>
                     <Col xs={10} className="pl-0">{trade.offer.paymentMethods.map(method => PAYMENT_METHODS[method]).join(', ')}</Col>
                   </Row>
                </CardBody>
                <CardFooter className="bg-white text-right border-0 pt-0 clickable">
                  <p className="m-0 border-top pt-2">
                    Buy <span className="text-black"><img
                    src={getTokenImage(trade.token.symbol)}
                    alt={trade.token.symbol + ' icon'}/> {trade.token.symbol}</span> at <span
                    className="font-weight-bold text-black">{truncateTwo(calculateEscrowPrice(trade, this.props.prices))} {trade.currency}</span>
                  </p>
                </CardFooter>
              </Card>);
            });
            if (trades.every(trade => trade === null)) {
              return (
                <Row className="my-1 border-bottom shadow-sm p-2 mb-3">
                  <Col className="align-self-center pt-3 pb-3 text-center text-muted">{this.props.t('trades.noFilteredTrades')}</Col>
                </Row>
              );
            }
            return trades;
          })()}
        </div>
      </Fragment>
    );
  }

  renderEmpty() {
    const {t, active} = this.props;
    return (
      <Card body className="text-center border-0 text-muted shadow-sm">
        {active ? t('trades.noActiveTrades') : t('trades.noPastTrades')}
      </Card>
    );
  }

  render() {
    const {t, trades} = this.props;
    return (
      <div className="mb-5">
        {trades.length === 0 ? this.renderEmpty(t) : this.renderTrades()}
      </div>
    );
  }
}

Trades.defaultProps = {
  active: false
};

Trades.propTypes = {
  t: PropTypes.func,
  trades: PropTypes.array,
  prices: PropTypes.object,
  address: PropTypes.string,
  tradeClick: PropTypes.func,
  active: PropTypes.bool
};

export default withNamespaces()(Trades);
