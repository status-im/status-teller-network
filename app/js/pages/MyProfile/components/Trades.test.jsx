import React from "react";
import { shallow } from "enzyme";

import Trades from "./Trades";

describe('Trades', () => {
  it('should render correctly', () => {
    const component = shallow(<Trades trades={[{offer: { owner: '0x0', token: { symbol: 'SNT' }, user: { username: 'Name' } }, value: '2', status: 'open'}]} />);
  
    expect(component).toMatchSnapshot();
  });

  it('should render when empty', () => {
    const component = shallow(<Trades trades={[]} />);
  
    expect(component).toMatchSnapshot();
  });
});
