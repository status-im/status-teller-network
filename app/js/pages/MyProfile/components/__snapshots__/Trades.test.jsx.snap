// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Trades should render correctly 1`] = `
<div
  className="mt-3"
>
  <div>
    <h3
      className="d-inline-block"
    />
    <span
      className="float-right"
    >
      <Link
        className="float-right"
        replace={false}
        to="/buy"
      >
         
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f061",
                "M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z",
              ],
              "iconName": "arrow-right",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          title=""
          transform={null}
        />
      </Link>
    </span>
  </div>
  <Card
    body={true}
    className="py-2 px-3 shadow-sm"
    tag="div"
  >
    <div
      className="d-flex my-1"
      key="0"
    >
      <span
        className="flex-fill align-self-center"
      >
        <Identicon
          className="align-middle rounded-circle"
          seed="0x0"
        />
        <span
          className="ml-2"
        >
          Name
        </span>
      </span>
      <span
        className="flex-fill align-self-center"
      >
         
        SNT
      </span>
      <span
        className="flex-fill align-self-center text-right text-success"
      >
        <FontAwesomeIcon
          border={false}
          className="mr-2"
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f111",
                "M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8z",
              ],
              "iconName": "circle",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          title=""
          transform={null}
        />
        open
      </span>
    </div>
  </Card>
</div>
`;

exports[`Trades should render when empty 1`] = `
<div
  className="mt-3"
>
  <div>
    <h3
      className="d-inline-block"
    />
    <span
      className="float-right"
    >
      <Link
        className="float-right"
        replace={false}
        to="/buy"
      >
         
        <FontAwesomeIcon
          border={false}
          className=""
          fixedWidth={false}
          flip={null}
          icon={
            Object {
              "icon": Array [
                448,
                512,
                Array [],
                "f061",
                "M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z",
              ],
              "iconName": "arrow-right",
              "prefix": "fas",
            }
          }
          inverse={false}
          listItem={false}
          mask={null}
          pull={null}
          pulse={false}
          rotation={null}
          size={null}
          spin={false}
          symbol={false}
          title=""
          transform={null}
        />
      </Link>
    </span>
  </div>
  <Card
    body={true}
    className="text-center"
    tag="div"
  />
</div>
`;
